### PR TITLE
Port six collection-related components to TypeScript

### PR DESCRIPTION
**Pull request by rafpaf**
Ports to TS:
* `BaseItemsTable`
* `BaseTableItem`
* `BulkActions`
* `CollectionContentView`
* `CollectionCopyEntityModal`
* `ItemsTable`
* some related test files

### CODE REVIEW

File: frontend/src/metabase/collections/components/CollectionCopyEntityModal.tsx [hunk 1]
```diff
@@ -38,34 +34,20 @@ function CollectionCopyEntityModal({
...
-        {/* A shallow-copied newEntityObject will not include `uncopied` */}
-        {newEntityObject.uncopied?.length > 0
-          ? t`Duplicated ${entityObject.model}, but couldn't duplicate some questions`
-          : t`Duplicated ${entityObject.model}`}
-        <Link className={cx(CS.link, CS.textBold, CS.ml1)} to={newEntityUrl}>
-          {t`See it`}
-        </Link>
-      </div>,
-      { icon: entityObject.model },
-    );
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/CollectionCopyEntityModal.tsx:67]: Nothing invokes this component with the triggerToast prop set to anything, so this seems to be dead code

File: frontend/src/metabase/collections/components/CollectionCopyEntityModal.tsx [hunk 2]
```diff
@@ -23,7 +19,7 @@ function mapStateToProps(state, props) {
   };
 }
 
-const getTitle = (entityObject, isShallowCopy) => {
+const getTitle = (entityObject: any, isShallowCopy: boolean) => {
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/CollectionCopyEntityModal.tsx:22]: `entityObject` was a conundrum - I've left it as any

File: frontend/src/metabase-types/api/bookmark.ts [hunk 3]
```diff
@@ -2,7 +2,13 @@ import type { CardId, CardType } from "./card";
...
 import type { DashboardId } from "./dashboard";
 
-export const BOOKMARK_TYPES = ["card", "collection", "dashboard"] as const;
+export const BOOKMARK_TYPES = [
+  "card",
+  "collection",
+  "dashboard",
+  "snippet",
+  "indexed-entity",
+] as const;
```
Comment (sloansparger) [inline frontend/src/metabase-types/api/bookmark.ts:11]: nit, not seeing how this relates to this refactor?

File: frontend/src/metabase/collections/components/BaseItemsTable.tsx [hunk 4]
```diff
@@ -1,45 +1,60 @@
...
 
-function SortableColumnHeader({
+type SortableColumnHeaderProps = {
+  name: string;
+  sortingOptions: SortingOptions;
+  onSortingOptionsChange: (newSortingOptions: SortingOptions) => void;
+  children?: ReactNode;
+} & Partial<HTMLAttributes<HTMLDivElement>>;
+
+export enum Sort {
```
Comment (sloansparger) [inline frontend/src/metabase/collections/components/BaseItemsTable.tsx:39]: my man 🥲❤️

File: frontend/src/metabase/collections/components/BaseTableItem.tsx [hunk 5]
```diff
@@ -58,16 +69,17 @@ export function BaseTableItem({
...
   }, [item, onToggleSelected]);
 
   const renderRow = useCallback(() => {
     const canSelect =
-      collection.can_write && typeof onToggleSelected === "function";
+      collection?.can_write && typeof onToggleSelected === "function";
 
-    const lastEditInfo = item["last-edit-info"];
+    const lastEditInfo =
+      "last-edit-info" in item ? item["last-edit-info"] : undefined;
```
Comment (sloansparger) [inline frontend/src/metabase/collections/components/BaseTableItem.tsx:82]: what does this do that the previous code didn't?

File: frontend/src/metabase/collections/components/BaseItemsTable.tsx [hunk 6]
```diff
@@ -157,10 +175,8 @@ function BaseItemsTable({
               <ColumnHeader>
                 <BulkSelectWrapper>
                   <CheckBox
-                    checked={selectedItems?.length > 0 || false}
-                    indeterminate={
-                      (selectedItems?.length > 0 && hasUnselected) || false
-                    }
+                    checked={!!selectedItems?.length}
+                    indeterminate={!!selectedItems?.length && !!hasUnselected}
```
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BaseItemsTable.tsx:179]:
```suggestion
                    indeterminate={!!selectedItems?.length && hasUnselected}
```

File: frontend/src/metabase/collections/components/BaseTableItem.tsx [hunk 7]
```diff
@@ -112,7 +124,7 @@ export function BaseTableItem({
               size={16}
               status={item.moderated_status}
             />
-            {item.description && (
+            {item.description?.trim() && (
```
Comment (sloansparger) [inline frontend/src/metabase/collections/components/BaseTableItem.tsx:127]: I think we shouldn't need to do this. We should trust that our inputs have been correctly sanitized. If they're not I think it's best to solve this problem elsewhere as we likely display the description in a dozen places.

File: frontend/src/metabase/collections/components/BulkActions.tsx [hunk 8]
```diff
@@ -68,35 +87,42 @@ function BulkActions({
...
             </ToastCard>
           </BulkActionsToast>
         )}
       </Transition>
-      {!_.isEmpty(selectedItems) && selectedAction === "copy" && (
-        <Modal onClose={onCloseModal}>
-          <CollectionCopyEntityModal
-            entityObject={selectedItems[0]}
+      {!!selectedItems &&
+        !_.isEmpty(selectedItems) &&
```
Comment (sloansparger) [inline frontend/src/metabase/collections/components/BulkActions.tsx:100]: we do this check twice (and I think I'm going to add 2 more today), could you please move this check to a common named variable.

**Review by sloansparger — APPROVED** [review]
great work! couple of comments. the trim feedback i'd like to see changed before letting this get merged.

**Comment by sloansparger** [comment]
Oh sorry, just noticed. You should backport this. I think a lot of changes can/could happen in these files, so it would be advisable to reduce future conflicts given you're maintaining the same functionality here.

File: frontend/src/metabase/collections/components/BaseItemsTable.tsx (hunk 4 above)
Comment (npfitz) [reply frontend/src/metabase/collections/components/BaseItemsTable.tsx:39]: Can we use `PropsWithChildren` here?

File: frontend/src/metabase-types/api/bookmark.ts (hunk 3 above)
Comment (rafpaf) [reply frontend/src/metabase-types/api/bookmark.ts:11]: If I don't add those, I get this:

```
frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:108:42 - error TS2345: Argument of type '"card" | "collection" | "dashboard" | "snippet" | "indexed-entity"' is not assignable to parameter of type '"card" | "collection" | "dashboard"'.
  Type '"snippet"' is not assignable to type '"card" | "collection" | "dashboard"'.

108     toggleBookmark?.(item.id.toString(), normalizeItemModel(item));
                                             ~~~~~~~~~~~~~~~~~~~~~~~~
```

File: frontend/src/metabase/collections/components/BaseTableItem.tsx (hunk 5 above)
Comment (rafpaf) [reply frontend/src/metabase/collections/components/BaseTableItem.tsx:82]: Good catch

File: frontend/src/metabase-types/api/bookmark.ts (hunk 3 above)
Comment (sloansparger) [reply frontend/src/metabase-types/api/bookmark.ts:11]: my b, carry on then

**Comment by sloansparger** [comment]
discussed with raffi that we can skip backporting at this time

**Pull request merged**